# Data Mining Project
### By:  Group 3: Elio Aybar, Cristal Garcia, Sunny Li, Matt Norgren

### Data Cleaning

In [1]:
#Packages to import
import os
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt
import datetime as dt

pd.set_option('display.max_columns', 100) # Setting pandas to display a N number of columns
pd.set_option('display.max_rows', 10) # Setting pandas to display a N number rows
pd.set_option('display.width', 1000)

from sklearn.ensemble import AdaBoostRegressor,GradientBoostingRegressor,VotingClassifier,BaggingRegressor,RandomForestRegressor
from sklearn.model_selection import train_test_split
#from hypopt import GridSearch
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LogisticRegression,Perceptron,SGDRegressor
from sklearn.preprocessing import StandardScaler


import warnings
warnings.filterwarnings('ignore')

In [2]:
### IMPORT SET

#Change working directory to match where the large files are stored
# it is not best practice to centrally store files this large - although
# git LFS should be explored

### CWD -  MAN
%cd "C:\Users\MAN\OneDrive - The University of Chicago\2020_Autumn\Data Mining Platforms [MSCA 31008]\Group_Project\"
## CWD - Cristal
#%cd "C:\Users\cgarc\OneDrive - The University of Chicago\GitHub\US_HousingData-\docs"
#cd
## CWD - Elio 
#%cd "C:\Users\Tipo\Desktop\GitHub\Project"
## CWD - Sunny
#%cd "/Users/sunny/Data mining/final project"


C:\Users\MAN\OneDrive - The University of Chicago\2020_Autumn\Data Mining Platforms [MSCA 31008]\Group_Project


In [100]:
### Establish Pandas Data Frames / Import

### Zillow Set- Zip 
zip_df = pd.read_csv('Zip_time_series.csv', index_col = 'Date')  ### Issues with NaN on import

## Public School Ratings 
PS_df = pd.read_csv('Public_Schools.csv')

## Starbucks Upto 2017 for 70 Countries
SBUX_df = pd.read_csv('directory.csv')

## Food Desert 
foodlookup_df = pd.read_csv('food_access_variable_lookup.csv')
food_df = pd.read_csv('food_access_research_atlas.csv') 

## Hospital Quality Care 
hospital_df = pd.read_csv('Hospital General Information.csv', encoding="ISO-8859-1")

## Target Up to 2017 for US 
TGT_df = pd.read_csv('target.csv', encoding="ISO-8859-1")

### Individual Set Cleaning (Preparing for Combination

In [4]:
################################
## Food Desert -- Cleaning CG  #
################################
food_df.shape

(72864, 147)

In [5]:
#Columns to drop
food_df.drop(['GroupQuartersFlag', 'NUMGQTRS', 'PCTGQTRS', 'LILATracts_Vehicle',  'LATractsVehicle_20', 'LAPOP1_10', 'LAPOP05_10', 'LAPOP1_20', 'LALOWI1_10', 'LALOWI05_10', 'LALOWI1_20', 'lapophalf', 'lapophalfshare', 'lalowihalf', 'lalowihalfshare', 'lakidshalf', 'lakidshalfshare', 'laseniorshalf', 'laseniorshalfshare', 'lawhitehalf', 'lawhitehalfshare', 'lablackhalf', 'lablackhalfshare', 'laasianhalf', 'laasianhalfshare'], axis = 1, inplace = True)
food_df.drop(['lanhopihalf', 'lanhopihalfshare', 'laaianhalf', 'laaianhalfshare', 'laomultirhalf', 'laomultirhalfshare', 'lahisphalf', 'lahisphalfshare', 'lahunvhalf', 'lahunvhalfshare', 'lasnaphalf', 'lasnaphalfshare', 'lapop1', 'lapop1share', 'lalowi1', 'lalowi1share', 'lakids1', 'lakids1share', 'laseniors1', 'laseniors1share', 'lawhite1', 'lawhite1share', 'lablack1', 'lablack1share', 'laasian1', 'laasian1share', 'lanhopi1', 'lanhopi1share', 'laaian1', 'laaian1share', 'laomultir1', 'laomultir1share', 'lahisp1', 'lahisp1share', 'lahunv1', 'lahunv1share'], axis = 1, inplace = True)
food_df.drop(['lasnap1', 'lasnap1share', 'lapop10', 'lapop10share', 'lalowi10', 'lalowi10share', 'lakids10', 'lakids10share', 'laseniors10', 'laseniors10share', 'lawhite10', 'lawhite10share', 'lablack10', 'lablack10share', 'laasian10', 'laasian10share', 'lanhopi10', 'lanhopi10share', 'laaian10', 'laaian10share', 'laomultir10', 'laomultir10share', 'lahisp10', 'lahisp10share', 'lahunv10', 'lahunv10share', 'lasnap10', 'lasnap10share', 'lapop20', 'lapop20share', 'lalowi20', 'lalowi20share', 'lakids20', 'lakids20share', 'laseniors20', 'laseniors20share'], axis = 1, inplace = True)
food_df.drop(['lawhite20', 'lawhite20share', 'lablack20', 'lablack20share', 'laasian20', 'laasian20share', 'lanhopi20', 'lanhopi20share', 'laaian20', 'laaian20share', 'laomultir20', 'laomultir20share', 'lahisp20', 'lahisp20share', 'lahunv20', 'lahunv20share'], axis = 1, inplace = True)
food_df.drop(['LILATracts_1And10', 'LILATracts_halfAnd10', 'LILATracts_1And20', 'HUNVFlag', 'lasnap20', 'lasnap20share', 'TractLOWI', 'TractKids', 'TractSeniors', 'TractWhite', 'TractBlack', 'TractAsian', 'TractNHOPI', 'TractAIAN', 'TractOMultir', 'TractHispanic', 'TractHUNV', 'TractSNAP'], axis = 1, inplace = True)
food_df.drop(['LAhalfand10', 'LA1and20', 'LATracts_half', 'LATracts1', 'LATracts10', 'LATracts20'], axis = 1, inplace = True)

In [6]:
#Drop Duplication rows
food_df.drop_duplicates(inplace= True)

In [7]:
#There are rows in the data with censustract numbers that no longer apply and the respective row has zero values. Zipcodes 
#couldn't be identified and were replaced with NA's. Drop these respective rows with NA's in zipcode because those aren't 
#functional censustract numbers and the rows also have 0's for more than 70% of columns
food_df = food_df[food_df['Zipcode'].notna()]

KeyError: 'Zipcode'

In [ ]:
food_df.shape

In [ ]:
#Check missing values
food_df.apply(lambda x: sum(x.isnull()))

In [ ]:
#Check unique values
food_df.apply(lambda x: len(x.unique()))

In [ ]:
food_df.info()

In [ ]:
import seaborn as sns
#Visually represent collinearity
import matplotlib.pyplot as plt
plt.figure(figsize=(10,8))
sns.heatmap(food_df.corr(),annot=True,cmap="YlGnBu")
plt.show()

In [ ]:
food_df.shape

In [ ]:
#Groupby Zipcode and find the average of features
food_df = food_df.groupby('Zipcode', as_index=False)['Urban', 'POP2010', 'OHU2010', 'LowIncomeTracts', 'PovertyRate', 'MedianFamilyIncome', 'LA1and10'].mean()

In [ ]:
food_df.apply(lambda x: len(x.unique()))

In [ ]:
food_df.head()

In [ ]:
food_df.shape

In [ ]:
####################################################################
## Public School Ratings (2014-2015 School years) -- Cleaning CG  ##
####################################################################

PS_df.shape

In [ ]:
#Change 'LEVEL_' to change to integer (1: elementary, 2: middle school, 3: high school, 4:PK-13, N=0: Not specified)
PS_df['LEVEL_'] = PS_df['LEVEL_'].apply({'N':0, '1':1, '2':2, '3':3, '4':4 }.get)

In [ ]:
#Drop ZIP4 because 43% of data missing and already have other location indicators; don't need Telephone, website, source, 
#'VAL_METHOD', 'NAICS_CODE'; 'NAICS_DESC' is generalized description of 'LEVEL'; SHELTER_ID doesn't have 76% of data available 

PS_df.drop(['X', 'Y', 'OBJECTID', 'NCESID', 'ZIP4', 'TELEPHONE', 'NAICS_CODE', 'NAICS_DESC', 'SOURCE', 'VAL_METHOD', 'VAL_DATE',  'WEBSITE', 'ST_GRADE', 'END_GRADE', 'DISTRICTID',  'SHELTER_ID'], axis = 1, inplace = True)

#Drop Puerto Rico location
PS_df = PS_df[PS_df.COUNTRY != 'PRI']

#Drop Country column since all inland
PS_df.drop(['COUNTRY'], axis = 1, inplace = True)

#Convert SOURCEDATE to actual datatime type
PS_df['SOURCEDATE'] = pd.to_datetime(PS_df['SOURCEDATE'])

In [ ]:
#Drop Duplication rows
PS_df.drop_duplicates(inplace= True)

In [ ]:
#Drop schools with 0 and negative enrolment
PS_df = PS_df[PS_df.ENROLLMENT > 0]

In [ ]:
#Drop schools with 0 and negative count of teachers
PS_df = PS_df[PS_df.FT_TEACHER > 0]

In [ ]:
#Create Student: Teacher ratio based on enrollment and Teacher count. 30:1 is the ideal ratio 
PS_df['Class_Teacher_RATIO'] = PS_df['ENROLLMENT'] / PS_df['FT_TEACHER']

In [ ]:
import seaborn as sns
#Visually represent collinearity
import matplotlib.pyplot as plt
plt.figure(figsize=(10,8))
sns.heatmap(PS_df.corr(),annot=True,cmap="YlGnBu")
plt.show()

In [ ]:
#Drop Population, Enrollment, and FT_Teacher columns since they have collinearity and enrollment and FT_teacher included in 
#Class_Teacher_RATIO
PS_df.drop(['POPULATION', 'ENROLLMENT', 'FT_TEACHER'], axis = 1, inplace = True)

In [ ]:
PS_df.head()

In [ ]:
PS_df.shape

In [ ]:
#Groupby Zipcode and find avg of features per zipcode
PS_df = PS_df.groupby('ZIP', as_index=False)['TYPE', 'STATUS', 'LEVEL_', 'Class_Teacher_RATIO'].mean()

In [ ]:
PS_df.apply(lambda x: len(x.unique()))

In [ ]:
PS_df.shape

In [ ]:
PS_df.head()

In [ ]:
##########################
## Target -- Cleaning EA #
##########################
TGT_df.shape

In [ ]:
TGT_df.info()

In [ ]:
#Select desired variables
TGT_df = TGT_df[['Address.FormattedAddress','Address.AddressLine1','Address.City','Address.CountryName','Address.Latitude','Address.Longitude','Address.County','Address.PostalCode','Address.Subdivision']]

In [ ]:
#Analize variable information
TGT_df.info()

In [ ]:
# Count the number of nan values in each column
print(TGT_df.isnull().sum())

In [ ]:
#Change variable names
TGT_df = TGT_df.rename(columns={"Address.FormattedAddress": "FormattedAddress", 
                   "Address.AddressLine1": "AddressLine1",
                   "Address.City": "City",
                   "Address.CountryName": "CountryName",
                   "Address.Latitude": "Latitude",
                   "Address.Longitude": "Longitude",
                   "Address.County": "County",
                   "Address.PostalCode": "PostalCode",
                   "Address.Subdivision": "Subdivision",
                  })

In [ ]:
#Drop missing values in column County variable
TGT_df = TGT_df[TGT_df.County.notnull()]

In [ ]:
TGT_df.head()

In [ ]:
# Count the number of nan values in each column
print(TGT_df.isnull().sum())

In [ ]:
TGT_df.head()

In [ ]:
#Add column that identifies Target at specified address (will groupby zipcode and count the number of targets by zipcode)
TGT_df['Target_count'] = 1
TGT_df = TGT_df.groupby('PostalCode', as_index=False)['Target_count'].count()

In [ ]:
#Pull shortened zipcode for Target data
TGT_df['PostalCode'] = [x[:5] for x in TGT_df['PostalCode']]

In [ ]:
#Convert values to integers
TGT_df['PostalCode'] = TGT_df.PostalCode.astype(int)

In [ ]:
TGT_df.info()

In [ ]:
TGT_df.shape

In [ ]:
#############################
## Starbucks -- Cleaning EA #
#############################
SBUX_df.shape

In [ ]:
SBUX_df.info()

In [ ]:
A_1 = SBUX_df['Brand']
print(A_1.groupby(A_1).count())

In [ ]:
A_2 = SBUX_df['Country']
print(A_2.groupby(A_2).count())

In [ ]:
SBUX_df = SBUX_df[SBUX_df["Brand"] == 'Starbucks']

In [ ]:
SBUX_df = SBUX_df[SBUX_df["Country"] == 'US']

In [ ]:
#Select desired variables
SBUX_df = SBUX_df[['Brand','Street Address','City','State/Province','Country','Postcode','Longitude','Latitude']]

In [ ]:
SBUX_df.head()

In [ ]:
# Count the number of nan values in each column
print(SBUX_df.isnull().sum())

In [ ]:
#Drop missing values in column County variable
SBUX_df = SBUX_df[SBUX_df.Postcode .notnull()]

In [ ]:
# Count the number of nan values in each column
print(SBUX_df.isnull().sum())

In [ ]:
SBUX_df.info()

In [ ]:
#Add column that identifies Starbucks at specified address (will groupby zipcode and count the number of targets by zipcode)
SBUX_df['Starbucks_count'] = 1

In [ ]:
SBUX_df.head()

In [ ]:
SBUX_df.shape

In [ ]:
SBUX_df.apply(lambda x: len(x.unique()))

In [ ]:
#Pull shortened zipcode for Starbucks data
SBUX_df['Postcode'] = [x[:5] for x in SBUX_df['Postcode']]

In [ ]:
SBUX_df = SBUX_df.groupby('Postcode', as_index=False)['Starbucks_count'].count()

In [ ]:
#Convert values to integers
SBUX_df['Postcode']= SBUX_df.Postcode.astype(int)

In [ ]:
SBUX_df.info()

In [ ]:
SBUX_df.shape

In [ ]:
SBUX_df.head()

In [81]:
#######################
## Zip -- Cleaning MN #
#######################
zip_df.shape
#Understand range of outputs
#zip_df.describe()

(4383885, 75)

In [82]:
## Understand the number of nulls
zip_df.isnull().sum()

RegionName                                    0
InventorySeasonallyAdjusted_AllHomes    3332781
InventoryRaw_AllHomes                   3332781
MedianListingPricePerSqft_1Bedroom      4375691
MedianListingPricePerSqft_2Bedroom      4297331
                                         ...   
ZRI_AllHomes                            3046523
ZRI_AllHomesPlusMultifamily             3044532
ZriPerSqft_AllHomes                     3129916
Zri_MultiFamilyResidenceRental          3660343
Zri_SingleFamilyResidenceRental         3049564
Length: 75, dtype: int64

In [101]:
## Establish two new columns in order to test logic of replacement / consolidation
zip_df['NumberOfBedrooms'] = 0
zip_df['DwellingType'] = 0
zip_df['MedianListingPricePerSqft'] = 0
zip_df['MedianListingPrice'] = 0
zip_df['MedianRentalPrice'] = 0
zip_df['MedianRentalPricePerSqft'] = 0
zip_df['MedianListPPS_D'] = 0
zip_df['MedianList_D'] = 0
zip_df['MedianRentalPPS_D'] = 0
zip_df['MedianRental_D'] = 0

In [102]:
## Replace null with 0
zip_df['MedianListingPricePerSqft_1Bedroom'] = zip_df['MedianListingPricePerSqft_1Bedroom'].fillna(0)
zip_df['MedianListingPricePerSqft_2Bedroom'] = zip_df['MedianListingPricePerSqft_2Bedroom'].fillna(0)
zip_df['MedianListingPricePerSqft_3Bedroom'] = zip_df['MedianListingPricePerSqft_3Bedroom'].fillna(0)
zip_df['MedianListingPricePerSqft_4Bedroom'] = zip_df['MedianListingPricePerSqft_4Bedroom'].fillna(0)
zip_df['MedianListingPricePerSqft_5BedroomOrMore'] = zip_df['MedianListingPricePerSqft_5BedroomOrMore'].fillna(0)
##
zip_df['MedianListingPrice_1Bedroom'] = zip_df['MedianListingPrice_1Bedroom'].fillna(0)
zip_df['MedianListingPrice_2Bedroom'] = zip_df['MedianListingPrice_2Bedroom'].fillna(0)
zip_df['MedianListingPrice_3Bedroom'] = zip_df['MedianListingPrice_3Bedroom'].fillna(0)
zip_df['MedianListingPrice_4Bedroom'] = zip_df['MedianListingPrice_4Bedroom'].fillna(0)
zip_df['MedianListingPrice_5BedroomOrMore'] = zip_df['MedianListingPrice_5BedroomOrMore'].fillna(0)
##
zip_df['MedianRentalPricePerSqft_1Bedroom'] = zip_df['MedianRentalPricePerSqft_1Bedroom'].fillna(0)
zip_df['MedianRentalPricePerSqft_2Bedroom'] = zip_df['MedianRentalPricePerSqft_2Bedroom'].fillna(0)
zip_df['MedianRentalPricePerSqft_3Bedroom'] = zip_df['MedianRentalPricePerSqft_3Bedroom'].fillna(0)
zip_df['MedianRentalPricePerSqft_4Bedroom'] = zip_df['MedianRentalPricePerSqft_4Bedroom'].fillna(0)
zip_df['MedianRentalPricePerSqft_5BedroomOrMore'] = zip_df['MedianRentalPricePerSqft_5BedroomOrMore'].fillna(0)
##
zip_df['MedianRentalPrice_1Bedroom'] = zip_df['MedianRentalPrice_1Bedroom'].fillna(0)
zip_df['MedianRentalPrice_2Bedroom'] = zip_df['MedianRentalPrice_2Bedroom'].fillna(0)
zip_df['MedianRentalPrice_3Bedroom'] = zip_df['MedianRentalPrice_3Bedroom'].fillna(0)
zip_df['MedianRentalPrice_4Bedroom'] = zip_df['MedianRentalPrice_4Bedroom'].fillna(0)
zip_df['MedianRentalPrice_5BedroomOrMore'] = zip_df['MedianRentalPrice_5BedroomOrMore'].fillna(0)
##
zip_df['MedianListingPrice_AllHomes'] = zip_df['MedianListingPrice_AllHomes'].fillna(0)
zip_df['MedianListingPrice_CondoCoop'] = zip_df['MedianListingPrice_CondoCoop'].fillna(0)
zip_df['MedianListingPrice_DuplexTriplex'] = zip_df['MedianListingPrice_DuplexTriplex'].fillna(0)
zip_df['MedianListingPrice_SingleFamilyResidence'] = zip_df['MedianListingPrice_SingleFamilyResidence'].fillna(0)
##
zip_df['MedianListingPrice_AllHomes'] = zip_df['MedianListingPrice_AllHomes'].fillna(0)
zip_df['MedianListingPrice_CondoCoop'] = zip_df['MedianListingPrice_CondoCoop'].fillna(0)
zip_df['MedianListingPrice_DuplexTriplex'] = zip_df['MedianListingPrice_DuplexTriplex'].fillna(0)
zip_df['MedianListingPrice_SingleFamilyResidence'] = zip_df['MedianListingPrice_SingleFamilyResidence'].fillna(0)
##
zip_df['MedianListingPricePerSqft_AllHomes'] = zip_df['MedianListingPricePerSqft_AllHomes'].fillna(0)
zip_df['MedianListingPricePerSqft_CondoCoop'] = zip_df['MedianListingPricePerSqft_CondoCoop'].fillna(0)
zip_df['MedianListingPricePerSqft_DuplexTriplex'] = zip_df['MedianListingPricePerSqft_DuplexTriplex'].fillna(0)
zip_df['MedianListingPricePerSqft_SingleFamilyResidence'] = zip_df['MedianListingPricePerSqft_SingleFamilyResidence'].fillna(0)
##
zip_df['MedianRentalPrice_AllHomes'] = zip_df['MedianRentalPrice_AllHomes'].fillna(0)
zip_df['MedianRentalPrice_CondoCoop'] = zip_df['MedianRentalPrice_CondoCoop'].fillna(0)
zip_df['MedianRentalPrice_DuplexTriplex'] = zip_df['MedianRentalPrice_DuplexTriplex'].fillna(0)
zip_df['MedianRentalPrice_SingleFamilyResidence'] = zip_df['MedianRentalPrice_SingleFamilyResidence'].fillna(0)
##
zip_df['MedianRentalPricePerSqft_AllHomes'] = zip_df['MedianRentalPricePerSqft_AllHomes'].fillna(0)
zip_df['MedianRentalPricePerSqft_CondoCoop'] = zip_df['MedianRentalPricePerSqft_CondoCoop'].fillna(0)
zip_df['MedianRentalPricePerSqft_DuplexTriplex'] = zip_df['MedianRentalPricePerSqft_DuplexTriplex'].fillna(0)
zip_df['MedianRentalPricePerSqft_SingleFamilyResidence'] = zip_df['MedianRentalPricePerSqft_SingleFamilyResidence'].fillna(0)
zip_df['MedianRentalPricePerSqft_MultiFamilyResidence5PlusUnits'] = zip_df['MedianRentalPricePerSqft_MultiFamilyResidence5PlusUnits'].fillna(0)
zip_df['MedianRentalPricePerSqft_Studio'] = zip_df['MedianRentalPricePerSqft_Studio'].fillna(0)
zip_df['MedianPctOfPriceReduction_AllHomes'] = zip_df['MedianPctOfPriceReduction_AllHomes'].fillna(0)
zip_df['MedianPctOfPriceReduction_CondoCoop'] = zip_df['MedianPctOfPriceReduction_CondoCoop'].fillna(0)
zip_df['MedianPctOfPriceReduction_SingleFamilyResidence'] = zip_df['MedianPctOfPriceReduction_SingleFamilyResidence'].fillna(0)
zip_df['MedianPriceCutDollar_AllHomes'] = zip_df['MedianPriceCutDollar_AllHomes'].fillna(0)
zip_df['MedianPriceCutDollar_CondoCoop'] = zip_df['MedianPriceCutDollar_CondoCoop'].fillna(0)
zip_df['MedianPriceCutDollar_SingleFamilyResidence'] = zip_df['MedianPriceCutDollar_SingleFamilyResidence'].fillna(0)
zip_df['MedianRentalPricePerSqft_MultiFamilyResidence5PlusUnits'] = zip_df['MedianRentalPricePerSqft_MultiFamilyResidence5PlusUnits'].fillna(0)
zip_df['MedianRentalPricePerSqft_Studio'] = zip_df['MedianRentalPricePerSqft_Studio'].fillna(0)

In [103]:
## Optimized set in order to iterate through and assign a value to the new column created
## Many prior attempts utilized a more if then style structure but inevitably took a considerably long time
## and replaced when the value was 0  thus overwriting the prior data at every run
##WC: https://guillim.github.io/pandas/2018/10/22/Pandas-if-else-on-columns.html 

condlist = [
    (zip_df['MedianListingPricePerSqft_1Bedroom'] != 0)
    ,(zip_df['MedianListingPricePerSqft_2Bedroom'] != 0)
    ,(zip_df['MedianListingPricePerSqft_3Bedroom'] != 0)
    ,(zip_df['MedianListingPricePerSqft_4Bedroom'] != 0)
    ,(zip_df['MedianListingPricePerSqft_5BedroomOrMore'] != 0)
]

condlist2 = [(zip_df['MedianRentalPricePerSqft_AllHomes'] != 0)
    ,(zip_df['MedianRentalPricePerSqft_CondoCoop'] != 0)
    ,(zip_df['MedianRentalPricePerSqft_DuplexTriplex'] != 0)
    ,(zip_df['MedianRentalPricePerSqft_SingleFamilyResidence'] != 0)
    ,(zip_df['MedianRentalPricePerSqft_MultiFamilyResidence5PlusUnits'] != 0)
    ,(zip_df['MedianRentalPricePerSqft_Studio'] != 0)
]
      
#Establish if the criteria is met - what is the result
choicelist = [1,2,3,4,5]
choicelist2 = ['All','Condo','Duplex+','Single','Multi5+','Studio']
#Overwrite the number of bedroom columns with the given logic output 
zip_df['NumberOfBedrooms'] = np.select(condlist,choicelist)
zip_df['DwellingType'] = np.select(condlist2,choicelist2)

In [104]:
## Count per of decided bedrooms
zip_df['NumberOfBedrooms'].groupby(zip_df['NumberOfBedrooms']).count()

NumberOfBedrooms
0    3951999
1       8194
2      80569
3     310476
4      30627
5       2020
Name: NumberOfBedrooms, dtype: int64

In [105]:
## Count per of decided dwelling types
zip_df['DwellingType'].groupby(zip_df['DwellingType']).count()

DwellingType
0          4234042
All         120843
Condo          677
Duplex+       1231
Multi5+      25631
Single         924
Studio         537
Name: DwellingType, dtype: int64

In [ ]:
##Further Cleanup of like variables

## Optimized set in order to iterate through and assign a value to the new column created
## Many prior attempts utilized a more if then style structure but inevitably took a considerably long time
## and replaced when the value was 0  thus overwriting the prior data at every run
##WC: https://guillim.github.io/pandas/2018/10/22/Pandas-if-else-on-columns.html 

#condlist2 = [
#    (zip_df['MedianListingPrice_1Bedroom'] != 0)
#    ,(zip_df['MedianRentalPrice_1Bedroom'] != 0)
#    ,(zip_df['MedianRentalPricePerSqft_1Bedroom'] != 0)
#    ,(zip_df['MedianListingPrice_2Bedroom'] != 0)
#    ,(zip_df['MedianRentalPrice_2Bedroom'] != 0)
#    ,(zip_df['MedianRentalPricePerSqft_2Bedroom'] != 0)
#    ,(zip_df['MedianListingPrice_3Bedroom'] != 0)
#    ,(zip_df['MedianRentalPrice_3Bedroom'] != 0)
#    ,(zip_df['MedianRentalPricePerSqft_3Bedroom'] != 0)
#    ,(zip_df['MedianListingPrice_4Bedroom'] != 0)
#    ,(zip_df['MedianRentalPrice_4Bedroom'] != 0)
#    ,(zip_df['MedianRentalPricePerSqft_4Bedroom'] != 0)
#    ,(zip_df['MedianListingPrice_5BedroomOrMore'] != 0)
#    ,(zip_df['MedianRentalPrice_5BedroomOrMore'] != 0)
#    ,(zip_df['MedianRentalPricePerSqft_5BedroomOrMore'] != 0)
#]
#Establish if the criteria is met - what is the result
#choicelist2 = [1,1,1,2,2,2,3,3,3,4,4,4,5,5,5]
#Overwrite the number of bedroom columns with the given logic output 
#zip_df['NumberOfBedrooms'] = np.select(condlist2,choicelist2)

In [106]:
##Combine the bedroom columns into the median listing price per sqft column to reduce column space

zip_df['MedianListingPricePerSqft'] = (zip_df['MedianListingPricePerSqft_1Bedroom']
          + zip_df['MedianListingPricePerSqft_2Bedroom']
          + zip_df['MedianListingPricePerSqft_3Bedroom']
          + zip_df['MedianListingPricePerSqft_4Bedroom']
          + zip_df['MedianListingPricePerSqft_5BedroomOrMore'])
zip_df['MedianListingPrice'] = (zip_df['MedianListingPrice_1Bedroom']
          + zip_df['MedianListingPrice_2Bedroom']
          + zip_df['MedianListingPrice_3Bedroom']
          + zip_df['MedianListingPrice_4Bedroom']
          + zip_df['MedianListingPrice_5BedroomOrMore'])
zip_df['MedianRentalPricePerSqft'] = (zip_df['MedianRentalPricePerSqft_1Bedroom']
          + zip_df['MedianRentalPricePerSqft_2Bedroom']
          + zip_df['MedianRentalPricePerSqft_3Bedroom']
          + zip_df['MedianRentalPricePerSqft_4Bedroom']
          + zip_df['MedianRentalPricePerSqft_5BedroomOrMore'])
zip_df['MedianRentalPrice'] = (zip_df['MedianRentalPrice_1Bedroom']
          + zip_df['MedianRentalPrice_2Bedroom']
          + zip_df['MedianRentalPrice_3Bedroom']
          + zip_df['MedianRentalPrice_4Bedroom']
          + zip_df['MedianRentalPrice_5BedroomOrMore'])
zip_df['MedianListPPS_D'] = (zip_df['MedianListingPricePerSqft_AllHomes']
          + zip_df['MedianListingPricePerSqft_CondoCoop']
          + zip_df['MedianListingPricePerSqft_DuplexTriplex']
          + zip_df['MedianListingPricePerSqft_SingleFamilyResidence'])
zip_df['MedianList_D'] = (zip_df['MedianListingPrice_AllHomes']
          + zip_df['MedianListingPrice_CondoCoop']
          + zip_df['MedianListingPrice_DuplexTriplex']
          + zip_df['MedianListingPrice_SingleFamilyResidence'])
zip_df['MedianRentalPPS_D'] = (zip_df['MedianRentalPricePerSqft_AllHomes']
          + zip_df['MedianRentalPricePerSqft_CondoCoop']
          + zip_df['MedianRentalPricePerSqft_DuplexTriplex']
          + zip_df['MedianRentalPricePerSqft_SingleFamilyResidence'])
zip_df['MedianRental_D'] = (zip_df['MedianRentalPrice_AllHomes']
          + zip_df['MedianRentalPrice_CondoCoop']
          + zip_df['MedianRentalPrice_DuplexTriplex']
          + zip_df['MedianRentalPrice_SingleFamilyResidence'])

In [107]:
### Drop unneeded columns
zip_df = zip_df[[
         'RegionName'
        , 'NumberOfBedrooms'
        , 'DwellingType'
        , 'MedianListingPricePerSqft'
        , 'MedianListingPrice'
        , 'MedianRentalPrice'
        , 'MedianRentalPricePerSqft'
        , 'MedianListPPS_D'
        , 'MedianList_D'
        , 'MedianRentalPPS_D'
        , 'MedianRental_D']]

In [108]:
## Factorize the dwelling
## Ensure weight is given to the size
Dwelling_Type = {'All':6,
             'Multi5':5,
             'Duplex+':4,
             'Condo':3,
             'Single':2,
             'Studio':1}
zip_df['DwellingType'] = zip_df['DwellingType'].map(Dwelling_Type)


In [109]:
## Create column for input / remove all 0s
zip_df['Test'] = (zip_df['NumberOfBedrooms']
+ zip_df['DwellingType']
+ zip_df['MedianListingPricePerSqft']
+ zip_df['MedianListingPrice']
+ zip_df['MedianRentalPrice']
+ zip_df['MedianRentalPricePerSqft']
+ zip_df['MedianListPPS_D']
+ zip_df['MedianList_D']
+ zip_df['MedianRentalPPS_D']
+ zip_df['MedianRental_D'])

zip_df['Test'] = zip_df['Test'].fillna(0)

In [110]:
zip_df = zip_df[(zip_df['Test'] !=0)]

In [111]:
zip_df.shape

(124212, 12)

In [90]:
## Establish date range to 2017 for normalization
#zip_df = zip_df.loc['2017-01-01':'2017-12-31']

In [112]:
## Create a groupby and average the inputs
a = zip_df.groupby('RegionName')['MedianListingPricePerSqft'
                               ,'MedianListingPrice'
                              , 'MedianRentalPrice'
                              , 'MedianRentalPricePerSqft'
                              , 'MedianListPPS_D'
                              , 'MedianList_D'
                              , 'MedianRentalPPS_D'
                              , 'MedianRental_D'].mean()

In [113]:
a

MedianListingPricePerSqft  MedianListingPrice  MedianRentalPrice  MedianRentalPricePerSqft  MedianListPPS_D  MedianList_D  MedianRentalPPS_D  MedianRental_D
RegionName                                                                                                                                                              
1545                       714.050840        1.288503e+06        1618.875000                  0.000000       473.898823  8.374666e+05           1.526152        0.000000
2108                         0.000000        0.000000e+00        5861.666667                  7.825665         0.000000  0.000000e+00           8.683647     6002.083333
2109                         0.000000        0.000000e+00        1264.098361                  7.327604         0.000000  0.000000e+00           7.128927     4385.909836
2110                         0.000000        0.000000e+00        3462.264706                  2.961326         0.000000  0.000000e+00           7.958324     5714.705882
2111                         0.000000        0.000000e+00        3773.429577                  6.969672       831.690030  0.000000e+00           7.497222     4899.021127
...                               ...                 ...                ...                       ...              ...           ...                ...             ...
99507                      541.365317        8.486500e+05        1313.625000                  0.000000       539.038921  8.376996e+05           1.346605     1664.791667
99508                      460.585147        7.066792e+05        1047.083333                  1.251315       572.727445  1.025350e+06           1.268009        0.000000
99515                      370.869951        7.505041e+05           0.000000                  0.000000       547.015116  9.578478e+05           1.284740     1570.625000
99577                      352.399053        7.193307e+05         394.107143                  0.251335       453.252864  8.627969e+05           2.266878     3847.991071
99654                      447.826836        1.123318e+06        1389.583333                  0.000000       372.519377  9.941854e+05           1.015827     2730.750000

[2697 rows x 8 columns]

In [114]:
zipt = zip_df

In [115]:
zipt_df = pd.merge(zipt,a,how='left',on='RegionName')
zipt_df

RegionName  NumberOfBedrooms  DwellingType  MedianListingPricePerSqft_x  MedianListingPrice_x  MedianRentalPrice_x  MedianRentalPricePerSqft_x  MedianListPPS_D_x  MedianList_D_x  MedianRentalPPS_D_x  MedianRental_D_x          Test  MedianListingPricePerSqft_y  MedianListingPrice_y  MedianRentalPrice_y  MedianRentalPricePerSqft_y  MedianListPPS_D_y  MedianList_D_y  MedianRentalPPS_D_y  MedianRental_D_y
0            23462                 2           6.0                   393.132583              436950.0                  0.0                    0.000000         392.302637        369800.0             1.727785            2420.0  8.099652e+05                   362.403085          5.268924e+05          1966.015789                    1.452203         364.341815    4.817400e+05             2.589508       3683.457895
1            32082                 3           6.0                   472.948762                   0.0                  0.0                    0.000000         568.464169             0.0             1.069519               0.0  1.051482e+03                   506.414389          5.428717e+05             0.000000                    0.000000         587.265390    7.128802e+05             3.360673          0.000000
2            32244                 0           6.0                     0.000000                   0.0                  0.0                    0.000000           0.000000             0.0             1.365058               0.0  7.365058e+00                   100.730222          1.333742e+05          1287.415789                    0.900400          80.323252    1.231275e+05             1.401322       2065.847368
3            32547                 3           6.0                   113.687612                   0.0                  0.0                    0.000000           0.000000             0.0             1.411395            1590.0  1.714099e+03                   183.885963          1.860926e+05          1190.300000                    0.864714         148.135659    2.002986e+05             1.593364       2061.736842
4            32907                 0           6.0                     0.000000                   0.0                  0.0                    0.000000           0.000000             0.0             1.206959               0.0  7.206959e+00                    92.212513          1.710355e+05           179.752632                    0.127562          93.328215    1.499525e+05             1.344397       1995.057895
...            ...               ...           ...                          ...                   ...                  ...                         ...                ...             ...                  ...               ...           ...                          ...                   ...                  ...                         ...                ...             ...                  ...               ...
124207       99507                 2           6.0                   531.858465              856150.0               1275.0                    0.000000         536.899850        819900.0             1.284652            1700.0  1.680103e+06                   541.365317          8.486500e+05          1313.625000                    0.000000         539.038921    8.376996e+05             1.346605       1664.791667
124208       99508                 2           6.0                   437.917546              662350.0               1000.0                    1.187500         563.442697        976400.0             1.209358               0.0  1.640762e+06                   460.585147          7.066792e+05          1047.083333                    1.251315         572.727445    1.025350e+06             1.268009          0.000000
124209       99515                 3           6.0                   365.027071              749450.0                  0.0                    0.000000         543.203108        977999.5             1.259349            1700.0  1.730068e+06                   370.869951          7.505041e+05  

In [116]:
zipt_df['RegionName'].groupby(zipt_df['RegionName']).count()

RegionName
1545     12
2108     12
2109     61
2110     17
2111     71
         ..
99507    12
99508    12
99515    12
99577    56
99654    12
Name: RegionName, Length: 2697, dtype: int64

In [117]:
zipt_df.describe()

RegionName  NumberOfBedrooms   DwellingType  MedianListingPricePerSqft_x  MedianListingPrice_x  MedianRentalPrice_x  MedianRentalPricePerSqft_x  MedianListPPS_D_x  MedianList_D_x  MedianRentalPPS_D_x  MedianRental_D_x          Test  MedianListingPricePerSqft_y  MedianListingPrice_y  MedianRentalPrice_y  MedianRentalPricePerSqft_y  MedianListPPS_D_y  MedianList_D_y  MedianRentalPPS_D_y  MedianRental_D_y
count  124212.000000     124212.000000  124212.000000                124212.000000          1.242120e+05        124212.000000               124212.000000      124212.000000    1.242120e+05        124212.000000     124212.000000  1.242120e+05                124212.000000          1.242120e+05        124212.000000               124212.000000      124212.000000    1.242120e+05        124212.000000     124212.000000
mean    53175.761601          1.798522       5.912456                   280.352588          4.810137e+05          1326.630112                    1.004996         405.085952    5.780796e+05             2.281163       2738.767704  1.063855e+06                   280.352588          4.810137e+05          1326.630112                    1.004996         405.085952    5.780796e+05             2.281163       2738.767704
std     29225.308532          1.369154       0.555101                   461.326473          9.722319e+05          2560.653039                    1.846182         503.803183    7.816875e+05             2.212641       2273.561992  1.585187e+06                   390.267055          7.662740e+05          2225.117237                    1.614245         470.511133    6.489572e+05             2.069261       2019.805004
min      1545.000000          0.000000       1.000000                     0.000000          0.000000e+00             0.000000                    0.000000           0.000000    0.000000e+00             0.000000          0.000000  1.000000e+00                     0.000000          0.000000e+00             0.000000                    0.000000           0.000000    0.000000e+00             0.000000          0.000000
25%     30265.000000          0.000000       6.000000                     0.000000          0.000000e+00             0.000000                    0.000000         169.343483    6.278750e+04             1.352885       1600.000000  2.325854e+05                    47.263248          4.625012e+04             0.000000                    0.000000         172.352492    2.493086e+05             1.381902       1703.228916
50%     37130.000000          2.000000       6.000000                   181.753789          2.492250e+05             0.000000                    0.000000         254.731949    4.129000e+05             1.651134       2440.000000  7.078244e+05                   190.174743          3.123391e+05           518.012048                    0.317358         246.510884    4.018677e+05             1.654215       2422.291667
75%     85023.000000          3.000000       6.000000                   359.648248          6.214625e+05          1750.000000                    1.284872         462.790214    7.298912e+05             2.353627       3390.000000  1.355262e+06                   349.636157          6.186629e+05          1720.851190                    1.294411         453.252864    6.944449e+05             2.414155       3254.571429
max     99654.000000          5.000000       6.000000                  8471.512268          2.220000e+07         95000.000000                   36.422619        5467.523449    1.311650e+07            45.133745      80000.000000  3.353671e+07                  4876.228986          1.585665e+07         67175.000000                   29.423033        4893.946331    1.199592e+07            40.439551      72500.000000

In [ ]:
## Still need to factorize dwelling type



In [ ]:
#########################################
## Hospital Quality Care -- Cleaning SL #
#########################################
hospital_df.shape

In [ ]:
hospital_df.head()

In [ ]:
hospital_df.info()

In [ ]:
null_df = hospital_df.isnull().sum(axis = 0).sort_values().to_frame('missing_value')
null_df[null_df['missing_value'] > 0].reset_index()

In [ ]:
# drop provider id, hospital name, address, phone number since they are all irrelevant 
# also drop all footnote since they are all incorporate in the other columns.
drop_these = ['Provider ID', 'Hospital Name', 'Phone Number']
drop_these.extend([col for col in hospital_df if 'footnote' in col])
hospital_df = hospital_df.drop(columns = drop_these)

# hospital_df.loc[:,hospital_df.columns.str.endswith('footnote')]

In [ ]:
hospital_df.head()

In [ ]:
hospital_df.info()

In [ ]:
null_df = hospital_df.isnull().sum(axis = 0).sort_values().to_frame('missing_value')
null_df[null_df['missing_value'] > 0].reset_index()

In [ ]:
hospital_df['Meets criteria for meaningful use of EHRs'].value_counts()

In [ ]:
hospital_df['Meets criteria for meaningful use of EHRs'] = hospital_df['Meets criteria for meaningful use of EHRs'].fillna('Not Available')

In [ ]:
hospital_df['County Name'].value_counts()

In [ ]:
hospital_df.head()

In [ ]:
hospital_df['Hospital Ownership'].value_counts()

In [ ]:
hospital_df['Hospital Ownership'] = hospital_df['Hospital Ownership'].replace({
                    'Voluntary non-profit - Private': 'non-profit', 
                    'Proprietary':'for-profit', 
                    'Government - Hospital District or Authority': 'government',
                    'Voluntary non-profit - Other':'non-profit',
                    'Government - Local':'government',
                    'Voluntary non-profit - Church':'non-profit',
                    'Physician':'for-profit',
                    'Government - State':'government',
                    'Tribal ':'government',
                    'Government - Federal':'government',
                                  })

In [ ]:
hospital_df['Hospital Type'].value_counts()

There are four level in the following 7 aspect: 
Mortality national comparison; 
Safety of care national comparison;
Safety of care national comparison;
Readmission national comparison	Patient experience national comparison	Effectiveness of care national comparison	Timeliness of care national comparison	Efficient use of medical imaging national comparison
-1 Below the national average
0Same as the national average                    
1Above the national average       
Not Available   

In [ ]:
# there are four level in the following 7 aspect
#Same as the national average    
#Not Available                   
#Above the national average       
#Below the national average
hospital_df['Mortality national comparison'].value_counts()

In [ ]:
hospital_df['Safety of care national comparison'].value_counts()

In [ ]:

hospital_df['Readmission national comparison'].value_counts()

In [ ]:
hospital_df['Safety of care national comparison'].value_counts()

### Consolidation
https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html

In [ ]:
#### Consolidate dataframes for ease of use
#pd.concat()
PS_df.shape

In [ ]:
food_df.head()

In [ ]:
#Rename Column
food_df.rename(columns = {'Zipcode': 'ZIP'}, inplace=True)

In [ ]:
#Left Join food to PS data (want to make sure keep as much data as possible)
PS_food_df = PS_df.merge(food_df, on="ZIP", how="left") 
PS_food_df.head()

In [ ]:
PS_food_df.shape

In [ ]:
TGT_df.shape

In [ ]:
SBUX_df.shape

In [ ]:
#Rename SBUX and TGT columns
SBUX_df.rename(columns = {'Postcode': 'ZIP'}, inplace=True)
TGT_df.rename(columns = {'PostalCode': 'ZIP'}, inplace=True)

In [ ]:
#Left Join TGT to SBUX
SBUX_TGT_df = SBUX_df.merge(TGT_df, on="ZIP", how="left")
SBUX_TGT_df.head()

In [ ]:
SBUX_TGT_df.shape

In [ ]:
features_df = PS_food_df.merge(SBUX_TGT_df, on="ZIP", how="left")
features_df.head()

In [ ]:
features_df.shape

### Dimensional Reduction & Quick Analysis

### Feature Engineering
https://pypi.org/project/autoimpute/

### Model Training

In [ ]:
#(test as much algorithm as possible, possible models include Random Forest, Gradient Boosting, Ada-Boosting, Voting Classifier...)

### Hyperparameter Tuning

Use Randome search training for parameters found in following article: 
https://towardsdatascience.com/hyperparameter-tuning-the-random-forest-in-python-using-scikit-learn-28d2aa77dd74

#### Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(2, 20, num = 10)]
max_depth.append(None)

#### Minimum number of samples required at each leaf node
min_samples_leaf = [0.05, 0.10, 0.13, 0.15, 0.20, 0.26, 0.5]

#### Create the random grid
random_grid = {'max_depth': max_depth, 'min_samples_leaf': min_samples_leaf}
pprint(random_grid)

#### Use the random grid to search for best hyperparameters
#### First create the base model to tune
dt = DecisionTreeRegressor()

#### Random search of parameters, using 10 fold cross validation, 
dt_random = RandomizedSearchCV(estimator = dt, param_distributions = random_grid, cv = 10, random_state=3, n_jobs = -1)
#dt_random = RandomizedSearchCV(estimator = dt, param_distributions = random_grid, n_iter = 100, cv = 10, verbose=2, random_state=3, n_jobs = -1)

#### Fit the random search model
dt_random.fit(X_train, y_train)

#### See what best parameters from fitting the random search: 
dt_random.best_params_

### Model Accuracy Testing